tools/src/blockchain-analyser/Main.hs

```diff
@@ -0,0 +1,98 @@
+import           Pos.Block.Core     (Block, blockHeaderHash, getBlockHeader)
+import           Pos.Block.Types    (Undo)
+import           Pos.Core           (HasCoreConstants, giveStaticConsts)
+import           Pos.Core.Types     (HeaderHash)
+import           Pos.DB             (closeNodeDBs, openNodeDBs)
+import qualified Pos.DB.Block       as DB
+import qualified Pos.DB.DB          as DB
+import           Pos.Ssc.GodTossing (SscGodTossing)
+import           System.Directory   (canonicalizePath, doesDirectoryExist, getFileSize,
+                                     listDirectory, withCurrentDirectory)
+
+import           Options            (CLIOptions (..), getOptions)
+import           Rendering          (render, renderBlock, renderBlocks, renderHeader)
+import           Types              (BlockchainInspector, DBFolderStat,
+                                     initBlockchainAnalyser, prevBlock)
+
+import           Universum
+
+-- | Like Unix's `du -s`, but works across all the major platforms and
+-- returns the total number of bytes the directory occupies on disk.
+-- Recurses inside each directory it finds.
+du_s :: FilePath -> IO Integer
+du_s root = go 0 [root]
+  where
+    go !acc [] = return acc
+    go !acc (f:fs) = do
+        sz        <- getFileSize f
+        isDir     <- doesDirectoryExist f
+        extraDirs <- if isDir then ls f else return mempty
+        go (acc + sz) (extraDirs <> fs)
+
+ls :: FilePath -> IO [FilePath]
+ls f = withCurrentDirectory f ((listDirectory >=> mapM canonicalizePath) f)
+
+-- | Calculate the size of each folder.
+dbSizes :: FilePath -> IO [DBFolderStat]
+dbSizes root = do
+    parents <- ls root
+    forM (root : parents) $ \f -> (toText f,) <$> du_s f
+
+-- | Analyse the blockchain, printing useful statistics.
+analyseBlockchain :: HasCoreConstants => CLIOptions -> HeaderHash -> BlockchainInspector ()
+analyseBlockchain cli tip =
+    if incremental cli then do liftIO $ putText (renderHeader cli)
+                               analyseBlockchainEagerly cli tip
+                       else analyseBlockchainLazily cli tip
+
+-- | Tries to fetch a `Block` given its `HeaderHash`.
+fetchBlock :: HasCoreConstants => HeaderHash -> BlockchainInspector (Maybe (Block SscGodTossing))
+fetchBlock = DB.dbGetBlockSumDefault @SscGodTossing
+
+-- | Tries to fetch an `Undo` for the given `Block`.
+fetchUndo :: HasCoreConstants => Block SscGodTossing -> BlockchainInspector (Maybe Undo)
+fetchUndo block = DB.blkGetUndo @SscGodTossing (blockHeaderHash $ getBlockHeader block)
+
+-- | Analyse the blockchain lazily by rendering all the blocks at once, loading the whole
+-- blockchain into memory. This mode generates very nice-looking tables, but using it for
+-- big DBs might not be feasible.
+analyseBlockchainLazily :: HasCoreConstants => CLIOptions -> HeaderHash -> BlockchainInspector ()
+analyseBlockchainLazily cli currentTip = do
+    allBlocks <- go currentTip mempty
```

Comment:
There is `loadBlundsFromTipWhile` in `Pos.DB.DB` which does the same thing (except that it returns `NewestFirst`).

Reply:
I've tried to use this one, but the `condition` of `loadBlundsFromTipWhile` is a pure function whereas my terminating condition is whether I can fetch the next block or not, for which I need to run a monadic computation in the DB monad. Dunno if there is some other trick I can play, but otherwise I guess we'll have to live with the current implementation 😞

Reply:
@adinapoli-iohk just pass `const True`, it will load all blocks and will stop when it reaches the deepest genesis block. That's exactly what you need, AFAIU. A small advantage of using `loadBlundsFromTipWhile` is that there will be `DBMalformed` exception when there is no previous block in DB for any block different from the deepest genesis block (which indeed shouldn't happen for valid DB), while your version will ignore it silently.